openrr-apps/src/robot_config.rs

```diff
@@ -322,54 +666,149 @@ impl RobotConfig {
             );
         }
-        speakers
+        Ok(speakers)
     }
 
     fn create_raw_joint_trajectory_clients(
         &self,
+        plugins: &mut PluginMap,
     ) -> Result<HashMap<String, Arc<dyn JointTrajectoryClient>>, Error> {
-        let urdf_robot = if let Some(urdf_path) = self.openrr_clients_config.urdf_full_path() {
-            let urdf_robot = urdf_rs::utils::read_urdf_or_xacro(urdf_path)?;
-            Some(urdf_robot)
-        } else {
-            None
-        };
-
+        let urdf_viz_clients_configs: Vec<_> = self
+            .urdf_viz_clients_configs
+            .iter()
+            .filter(|c| {
+                self.joint_trajectory_clients
+                    .as_ref()
+                    .map_or(true, |v| v.contains(&c.name))
+            })
+            .cloned()
+            .collect();
+        #[cfg(feature = "ros")]
+        let ros_clients_configs: Vec<_> = self
+            .ros_clients_configs
+            .iter()
+            .filter(|c| {
+                self.joint_trajectory_clients
+                    .as_ref()
+                    .map_or(true, |v| v.contains(&c.name))
+            })
+            .cloned()
+            .collect();
+
+        let mut urdf_robot = None;
         #[cfg(not(feature = "ros"))]
-        let raw_joint_trajectory_clients = create_joint_trajectory_clients_lazy(
-            self.urdf_viz_clients_configs.clone(),
-            urdf_robot.as_ref(),
-        )?;
+        let use_urdf = !urdf_viz_clients_configs.is_empty();
         #[cfg(feature = "ros")]
-        let raw_joint_trajectory_clients = {
-            let mut clients = if self.urdf_viz_clients_configs.is_empty() {
-                HashMap::new()
-            } else {
-                create_joint_trajectory_clients_lazy(
-                    self.urdf_viz_clients_configs.clone(),
-                    urdf_robot.as_ref(),
-                )?
-            };
-            clients.extend(arci_ros::create_joint_trajectory_clients_lazy(
-                self.ros_clients_configs.clone(),
+        let use_urdf = !urdf_viz_clients_configs.is_empty() || !ros_clients_configs.is_empty();
+        if use_urdf {
+            if let Some(urdf_path) = self.openrr_clients_config.urdf_full_path() {
+                urdf_robot = Some(urdf_rs::utils::read_urdf_or_xacro(urdf_path)?);
+            }
+        }
+
+        let mut clients = if urdf_viz_clients_configs.is_empty() {
+            HashMap::new()
+        } else {
+            arci_urdf_viz::create_joint_trajectory_clients_lazy(
+                urdf_viz_clients_configs,
                 urdf_robot.as_ref(),
-            )?);
-            clients
+            )?
         };
-        Ok(raw_joint_trajectory_clients)
+
+        #[cfg(feature = "ros")]
+        clients.extend(arci_ros::create_joint_trajectory_clients_lazy(
+            ros_clients_configs,
+            urdf_robot.as_ref(),
+        )?);
+
+        for (plugin_name, config) in &self.plugins {
```

Comment:
I feel it's better to make this a function, like  `create_plugin_xxx_yyy` to reduce the lines and make it easy to understand.

Reply:
Around the code, L678 has a similar to this. It's nice if these are merged **if possible**. (I think it's not easy)